kbfstool/md_dump.go

```diff
@@ -0,0 +1,323 @@
+}
+
+func getRevision(ctx context.Context, config libkbfs.Config,
+	tlfID libkbfs.TlfID, branchID libkbfs.BranchID,
+	revisionPartStr string) (libkbfs.MetadataRevision, error) {
+	if len(revisionPartStr) == 0 || revisionPartStr == "latest" {
+		if branchID == libkbfs.NullBranchID {
+			irmd, err := config.MDOps().GetForTLF(ctx, tlfID)
+			if err != nil {
+				return libkbfs.MetadataRevisionUninitialized,
+					err
+			}
+			return irmd.Revision, nil
+		}
+
+		irmd, err := config.MDOps().GetUnmergedForTLF(
+			ctx, tlfID, branchID)
+		if err != nil {
+			return libkbfs.MetadataRevisionUninitialized, err
+		}
+		return irmd.Revision, nil
+	}
+
+	base := 10
+	revisionStr := revisionPartStr
+	if strings.HasPrefix(revisionStr, "0x") {
+		base = 16
+		revisionStr = strings.TrimPrefix(revisionStr, "0x")
+	}
+	u, err := strconv.ParseUint(revisionStr, base, 64)
+	if err != nil {
+		return libkbfs.MetadataRevisionUninitialized, err
+	}
+	return libkbfs.MetadataRevision(u), nil
+}
+
+func mdGet(ctx context.Context, config libkbfs.Config, tlfID libkbfs.TlfID,
+	branchID libkbfs.BranchID, rev libkbfs.MetadataRevision) (
+	libkbfs.ImmutableRootMetadata, error) {
+	var irmds []libkbfs.ImmutableRootMetadata
+	var err error
+	if branchID == libkbfs.NullBranchID {
+		irmds, err = config.MDOps().GetRange(ctx, tlfID, rev, rev)
+		if err != nil {
+			return libkbfs.ImmutableRootMetadata{}, err
+		}
+	} else {
+		irmds, err = config.MDOps().GetUnmergedRange(
+			ctx, tlfID, branchID, rev, rev)
+		if err != nil {
+			return libkbfs.ImmutableRootMetadata{}, err
+		}
+	}
+
+	if len(irmds) >= 1 {
+		return irmds[0], nil
+	}
+
+	return libkbfs.ImmutableRootMetadata{}, nil
+}
+
+func mdParseAndGet(ctx context.Context, config libkbfs.Config, input string) (
+	libkbfs.ImmutableRootMetadata, error) {
+	parts := mdGetRe.FindStringSubmatch(input)
+	if parts == nil {
+		return libkbfs.ImmutableRootMetadata{},
+			fmt.Errorf("Could not parse %q", input)
+	}
+
+	tlfPartStr := parts[1]
+	branchPartStr := parts[2]
+	revisionPartStr := parts[3]
+
+	tlfID, err := getTlfID(ctx, config, tlfPartStr)
+	if err != nil {
+		return libkbfs.ImmutableRootMetadata{}, err
+	}
+
+	branchID, err := getBranchID(ctx, config, tlfID, branchPartStr)
+	if err != nil {
+		return libkbfs.ImmutableRootMetadata{}, err
+	}
+
+	rev, err := getRevision(ctx, config, tlfID, branchID, revisionPartStr)
+	if err != nil {
+		return libkbfs.ImmutableRootMetadata{}, err
+	}
+
+	return mdGet(ctx, config, tlfID, branchID, rev)
+}
+
+func getUserString(
+	ctx context.Context, config libkbfs.Config, uid keybase1.UID) string {
+	username, _, err := config.KeybaseService().Resolve(
+		ctx, fmt.Sprintf("uid:%s", uid))
+	if err != nil {
+		printError("md dump", err)
+		return uid.String()
+	}
+	return fmt.Sprintf("%s (uid:%s)", username, uid)
+}
+
+func mdDumpOne(ctx context.Context, config libkbfs.Config,
+	rmd libkbfs.ImmutableRootMetadata) error {
+	mdID, err := config.Crypto().MakeMdID(&rmd.BareRootMetadata)
```

Comment:
Why not use `rmd.mdID`?  An immutable root metadata can't be made without one, right?